tests/tests/helpers.rs

```diff
@@ -249,3 +254,112 @@ pub async fn pipe_from_stream_and_drop<T: Serialize>(
 		}
 	}
 }
+
+#[allow(dead_code)]
+pub async fn socks_server_no_auth() -> SocketAddr {
+	let mut config = server::Config::default();
+	config.set_dns_resolve(false);
+	let config = std::sync::Arc::new(config);
+
+	let listener = tokio::net::TcpListener::bind("127.0.0.1:0").await.unwrap();
+	let proxy_addr = listener.local_addr().unwrap();
+
+	spawn_socks_server(listener, config).await;
+
+	proxy_addr
+}
+
+#[allow(dead_code)]
+pub async fn spawn_socks_server(listener: tokio::net::TcpListener, config: std::sync::Arc<server::Config>) {
+	let addr = listener.local_addr().unwrap();
+	tokio::spawn(async move {
+		loop {
+			let (stream, _) = listener.accept().await.unwrap();
+			let mut socks5_socket = server::Socks5Socket::new(stream, config.clone());
+			socks5_socket.set_reply_ip(addr.ip());
+
+			socks5_socket.upgrade_to_socks5().await.unwrap();
+		}
+	});
+}
+
+#[allow(dead_code)]
+pub async fn connect_over_socks_stream(server_addr: SocketAddr) -> Socks5Stream<TcpStream> {
+	let target_addr = server_addr.ip().to_string();
+	let target_port = server_addr.port();
+
+	let socks_server = socks_server_no_auth().await;
+
+	fast_socks5::client::Socks5Stream::connect(
+		socks_server,
+		target_addr,
+		target_port,
+		fast_socks5::client::Config::default(),
+	)
+	.await
+	.unwrap()
+}
+
+#[pin_project(project = DataStreamProj)]
+#[allow(dead_code)]
+pub enum DataStream<T: tokio::io::AsyncRead + tokio::io::AsyncWrite + std::marker::Unpin> {
+	Socks5(#[pin] Socks5Stream<T>),
+}
+
+impl<T: tokio::io::AsyncRead + tokio::io::AsyncWrite + std::marker::Unpin> AsyncRead for DataStream<T> {
```

Comment:
Nit: `T` doesn't need AsyncWrite bound?

Reply:
nah, it doesn't work the Socks5Stream has trait bound on the type itself to T: AsyncRead + AsyncWrite + Unpin.

Thus, these AsyncWrite can't be omitted even if it shouldn't be needed